Improve wallet rescan API

ryanofsky opened the pull request:
This PR is just refactoring. There are no changes in behavior. It implements a request from @TheBlueMatt in https://github.com/bitcoin/bitcoin/pull/9827#issuecomment-282123202 to make rescan calls use timestamps instead of block index pointers.

ryanofsky committed:
Add RescanFromTime method and use from rpcdump

No change in behavior.

ryanofsky committed:
Move birthday optimization out of ScanForWalletTransactions

This change has no effect on wallet behavior.

On wallet startup, the transaction scan avoids reading any blocks with
timestamps older than the wallet birthday (less than nTimeFirstKey -
TIMESTAMP_WINDOW). This block skipping code currently resides in
CWallet::ScanForWalletTransactions but it doesn't really belong there because
it makes the implementation unnecessarily fragile and hard to understand, and
it never has any effect except at startup (because all other callers do their
rescans based on timestamps other than, but always greater or equal to,
nTimeFirstKey).

jonasschnelli commented:
utACK 4182e6c89640f8745f0461f5ef610f343ef26296

ryanofsky commented:
Rebased 2de5c79de7536c0edbfa4e0cb0175a9b8608d720 -> 9bb66ab66013fc3024fad00bceeb22a330b5bc1b ([pr/scanclean.2](https://github.com/ryanofsky/bitcoin/commits/pr/scanclean.2) -> [pr/scanclean.3](https://github.com/ryanofsky/bitcoin/commits/pr/scanclean.3))

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -1459,6 +1463,34 @@ void CWalletTx::GetAmounts(std::list<COutputEntry>& listReceived,
 }
 
 /**
+ * Scan active chain for relevant transactions after importing keys. This should
```
Shouldnt the docs (here and above) be in the .h and not the .cpp?

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -1459,6 +1463,34 @@ void CWalletTx::GetAmounts(std::list<COutputEntry>& listReceived,
 /**
+ * Scan active chain for relevant transactions after importing keys. This should
+ * be called whenever new keys are added to the wallet, with the oldest key
+ * creation time minus TIMESTAMP_WINDOW.
```
Maybe move TIMESTAMP_WINDOW into an implicit thing that this function applies to the provided timestamp (and move its definition out of wallet.h)?

TheBlueMatt reviewed (commented):
utACK 9bb66ab66013fc3024fad00bceeb22a330b5bc1b

ryanofsky committed:
Make CWallet::RescanFromTime comment less ambiguous

ryanofsky committed:
Handle TIMESTAMP_WINDOW within CWallet::RescanFromTime

This way CWallet::RescanFromTime callers don't need to subtract
TIMESTAMP_WINDOW themselves.

This is pure refactoring, there is no change in behavior.

ryanofsky reviewed (commented):
Added 2 commits 9bb66ab66013fc3024fad00bceeb22a330b5bc1b -> deaf48b046e573f6774d19e74b58918ed777cb14 ([pr/scanclean.3](https://github.com/ryanofsky/bitcoin/commits/pr/scanclean.3) -> [pr/scanclean.5](https://github.com/ryanofsky/bitcoin/commits/pr/scanclean.5), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/scanclean.3...pr/scanclean.5))

laanwj commented:
utACK deaf48b.
